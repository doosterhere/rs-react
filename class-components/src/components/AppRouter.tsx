import { createBrowserRouter, RouterProvider } from 'react-router-dom';

import { ROUTES } from '../constants';

import { HomePage, Page404 } from '../pages';
import { ListItemDetailed } from './ListItemDetailed/ListItemDetailed';

const router = createBrowserRouter([
  {
    path: ROUTES.HOME,
    element: <HomePage />,
    errorElement: <Page404 />,
    children: [
      {
        path: ROUTES.DETAIL,
        element: <ListItemDetailed />,
      },
    ],
  },
]);

const AppRouter = () => <RouterProvider router={router} />;

export { AppRouter };
