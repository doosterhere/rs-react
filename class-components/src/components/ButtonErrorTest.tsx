import { useState } from 'react';

const ButtonErrorTest = () => {
  const [hasError, setHasError] = useState(false);

  const handleThrowError = () => {
    setHasError(true);
  };

  if (hasError) {
    throw new Error('Throwing the error while rendering');
  }

  return (
    <>
      <button type="button" className="button-error" onClick={handleThrowError}>
        Throw an error
      </button>
    </>
  );
};

export default ButtonErrorTest;
