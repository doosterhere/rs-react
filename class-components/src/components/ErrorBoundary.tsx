import { Component, ReactNode } from 'react';

interface IErrorBoundaryProps {
  children: ReactNode;
}

class ErrorBoundary extends Component<IErrorBoundaryProps> {
  state = {
    hasError: false,
    errorMessage: '',
  };

  static getDerivedStateFromError(error: Error) {
    return { hasError: true, errorMessage: error.message.toString() };
  }

  componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
    console.error(`Error in error boundary: ${error}`, `Component stack: ${errorInfo.componentStack}`);
  }

  reloadPage() {
    window.location.assign('/');
  }

  render() {
    return this.state.hasError ? (
      <div className="container" style={{ display: 'flex', alignItems: 'center' }}>
        <h1 data-testid="error-boundary">Something went wrong!</h1>
        <button onClick={this.reloadPage} style={{ width: '300px' }}>
          Return to main page
        </button>
      </div>
    ) : (
      this.props.children
    );
  }
}

export { ErrorBoundary };
