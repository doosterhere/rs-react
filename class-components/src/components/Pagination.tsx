import { FC } from 'react';
import { useSearchParams } from 'react-router-dom';

interface IPaginationProps {
  pageCount: number;
}

const Pagination: FC<IPaginationProps> = ({ pageCount }) => {
  const pages = Array.from({ length: pageCount }, (_, index) => index + 1);
  const [searchParams, setSearchParams] = useSearchParams();
  const activePage = searchParams.get('page');

  const handlePageClick = (page: number) => {
    setSearchParams({
      search: searchParams.get('search') || '',
      page: page.toString(),
    });
  };

  return (
    <div className="pagination">
      <div className="pagination-arrow">
        <span>&#10094;</span>
      </div>
      {pages.map(page => (
        <div
          className={
            page === Number(activePage)
              ? 'pagination-page active'
              : 'pagination-page'
          }
          key={page}
          onClick={() => handlePageClick(page)}
        >
          {page}
        </div>
      ))}
      <div className="pagination-arrow">
        <span>&#10095;</span>
      </div>
    </div>
  );
};

export default Pagination;
