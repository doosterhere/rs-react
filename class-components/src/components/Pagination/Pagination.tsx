'use client';
import { FC, useState, useMemo, useEffect } from 'react';
import { useRouter } from 'next/router';

import { clsx } from 'clsx';

import classes from './Pagination.module.css';

import { useTheme } from '../ThemeContext';

interface IPaginationProps {
  itemsCount: number;
}

const Pagination: FC<IPaginationProps> = ({ itemsCount }) => {
  const pagesCount = Math.ceil(itemsCount / 10);
  const pages = useMemo(() => Array.from({ length: pagesCount }, (_, index) => index + 1), [pagesCount]);
  const router = useRouter();
  const { query } = router;
  const [activePage, setActivePage] = useState('1');
  const { theme } = useTheme();

  useEffect(() => {
    if (query.page) {
      setActivePage(query.page.toString());
    } else {
      setActivePage('1');
    }
  }, [query.page]);

  const handlePageClick = (page: number) => {
    const search = query.search || '';
    setActivePage(page.toString());
    const newQuery = { ...query, search, page: page.toString() };
    router.push({ pathname: router.pathname, query: newQuery });
  };

  return (
    <div className={clsx(classes.pagination, classes[theme.value])}>
      {pages.map(page => (
        <div
          className={clsx(classes.page, page === Number(activePage) && classes.active)}
          key={page}
          onClick={() => handlePageClick(page)}
          role="button"
        >
          {page}
        </div>
      ))}
    </div>
  );
};

export { Pagination };
