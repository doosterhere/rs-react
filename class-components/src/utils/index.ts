import { renderWithRouter } from './test-utils';

export { renderWithRouter };
