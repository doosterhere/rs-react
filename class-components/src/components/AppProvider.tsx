'use client';

import { ReactNode, useRef } from 'react';
import { Provider } from 'react-redux';

import '../styles/global.css';

import { AppStore, configStore } from '../store/store';
import { ThemeProvider } from '../components/ThemeContext';
import { ErrorBoundary } from './ErrorBoundary';

function AppProvider({ children }: { children: ReactNode }) {
  const storeRef = useRef<AppStore>();

  if (!storeRef.current) {
    storeRef.current = configStore();
  }

  return (
    <Provider store={storeRef.current}>
      <ErrorBoundary>
        <ThemeProvider>{children}</ThemeProvider>
      </ErrorBoundary>
    </Provider>
  );
}

export { AppProvider };
