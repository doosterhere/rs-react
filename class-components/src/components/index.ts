import SearchBar from './SearchBar.tsx';
import List from './List.tsx';
import ListItem from './ListItem.tsx';
import Loader from './Loader.tsx';
import ErrorBoundary from './ErrorBoundary.tsx';
import ButtonErrorTest from './ButtonErrorTest.tsx';
import Fallback from './Fallback.tsx';
import Modal from './Modal.tsx';
import AppRouter from './AppRouter.tsx';
import Layout from './Layout.tsx';

export {
  SearchBar,
  List,
  ListItem,
  Loader,
  ErrorBoundary,
  ButtonErrorTest,
  Fallback,
  Modal,
  AppRouter,
  Layout,
};
