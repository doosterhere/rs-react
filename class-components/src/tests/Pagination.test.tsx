import { screen, fireEvent } from '@testing-library/react';

import { renderWithProvider } from '../utils';
import { Pagination } from '../components/Pagination';

jest.mock('next/navigation', () => ({
  usePathname: () => '/',
  useRouter: () => ({ push: jest.fn() }),
  useSearchParams: () => ({ get: () => '' }),
}));

describe('Pagination', () => {
  it('should renders correctly with correct numbers of buttons', () => {
    const itemsCount = 34;
    const buttonsCount = Math.ceil(itemsCount / 10);

    const { container } = renderWithProvider(<Pagination itemsCount={itemsCount} />);

    expect(container).toMatchSnapshot();

    const buttons = screen.queryAllByRole('button');

    expect(buttons).toHaveLength(Math.ceil(buttonsCount));
  });

  it('should correct sets "active" class name for the buttons', () => {
    renderWithProvider(<Pagination itemsCount={15} />);

    const buttons = screen.queryAllByRole('button');

    const button_1 = buttons[0];
    const button_2 = buttons[1];

    expect(button_1).toHaveClass('page active');
    expect(button_2).not.toHaveClass('active');

    fireEvent.click(button_2);

    expect(button_1).not.toHaveClass('active');
    expect(button_2).toHaveClass('page active');
  });
});
