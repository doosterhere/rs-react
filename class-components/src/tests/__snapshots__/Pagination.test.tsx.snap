// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should renders correctly with correct numbers of buttons 1`] = `
<div>
  <main
    class="main light"
  >
    <div
      class="pagination"
    >
      <div
        class="page active"
        role="button"
      >
        1
      </div>
      <div
        class="page"
        role="button"
      >
        2
      </div>
      <div
        class="page"
        role="button"
      >
        3
      </div>
      <div
        class="page"
        role="button"
      >
        4
      </div>
    </div>
  </main>
</div>
`;
